integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/common/BenchUtils.scala

```diff
@@ -412,6 +532,33 @@ object BenchUtils {
   }
 }
 
+class BenchmarkListener(list: ListBuffer[SparkPlanNode]) extends QueryExecutionListener {
+
+  override def onSuccess(funcName: String, qe: QueryExecution, durationNs: Long): Unit = {
+    def toJson(plan: SparkPlan): SparkPlanNode = {
+      val children: Seq[SparkPlanNode] = plan match {
+        case s: AdaptiveSparkPlanExec => Seq(toJson(s.executedPlan))
+        case s: QueryStageExec => Seq(toJson(s.plan))
+        case _ => plan.children.map(child => toJson(child))
+      }
+      val metrics: Seq[SparkSQLMetric] = plan.metrics
+          .map(m => SparkSQLMetric(m._1, m._2.metricType, m._2.value)).toSeq
+
+      SparkPlanNode(
```

Comment:
if this takes the `plan` object, I think it means you don't have to match on `plan.nodeName` in other code. Instead it can be the actual class instance => better type safety.

Reply:
So this is the class that gets serialized to JSON using reflection via json4s, so we need to keep this to primitive types.

Reply:
The idea here is that data capture and report generation are typically separate steps (we want to be able generate new reports in the future based on data that we capture now).

Reply:
maybe we could move `skipWrappers` to the input stage, before JSON? But I understand now why it is this way.

Reply:
I was erring on the side of recording as much information as possible and then choosing what to report on later, but yes, we could choose to filter these wrappers out before saving the JSON. I'll try that out now and see how it looks.

Reply:
Speaking of metrics gathering, it would be great if we could grab task deserialization metrics as part of this as described in #831

Reply:
I filed a follow-up PR for the stage/task-level metrics: https://github.com/NVIDIA/spark-rapids/pull/847